File: backend/routers/recommend.py
from fastapi import APIRouter, HTTPException
from ..data.concerts import CONCERTS_DATA
from ..models.concert import RecommendRequest, RecommendResponse, RankedHotel, Hotel

router = APIRouter(prefix="/api/recommend", tags=["recommend"])

HOTEL_PRIORITY_LABELS = {
    "closest": "ใกล้ฮอลล์ที่สุด",
    "value": "คุ้มค่าที่สุด",
    "rated": "รีวิวดีที่สุด",
}


def _hotel_reasoning(hotel: dict, priorities: list[str], within_budget: bool) -> str:
    frags = []
    if "closest" in priorities:
        frags.append(f"อยู่ห่างจากฮอลล์แค่ {hotel['dist']}")
    if "value" in priorities:
        frags.append(f"ราคาย่อมเยาที่ ฿{hotel['price']:,}/คืน")
    if "rated" in priorities:
        frags.append(f"คะแนนรีวิวสูงถึง {hotel['rating']}/5")
    labels = " + ".join(HOTEL_PRIORITY_LABELS[p] for p in priorities if p in HOTEL_PRIORITY_LABELS)
    budget_note = "" if within_budget else " (ราคาเกินงบที่เหลือ)"
    return f'เพราะคุณเลือกเน้น "{labels}" — {hotel["name"]} ตรงเงื่อนไขคุณ เพราะ{" และ".join(frags)} ({hotel["reason"]}){budget_note}'


@router.post("", response_model=RecommendResponse)
def recommend(req: RecommendRequest):
    if req.concertId not in CONCERTS_DATA:
        raise HTTPException(status_code=404, detail="Concert not found")

    hotels = CONCERTS_DATA[req.concertId]["hotels"]
    priorities = req.hotelPriorities if req.hotelPriorities else ["closest"]

    # คำนวณงบที่เหลือสำหรับโรงแรม
    concert_prices = CONCERTS_DATA[req.concertId]["prices"]
    ticket_price = next((p["price"] for p in concert_prices if p["zone"] == req.preselectedZone), 0)
    other_costs = req.transportCost + req.merchCost + req.foodCost + req.otherCost
    hotel_budget_total = req.budget - ticket_price - other_costs
    hotel_budget_per_night = hotel_budget_total / max(req.hotelNights, 1)

    def hotel_score(h: dict) -> float:
        scores: dict[str, float] = {}
        if "closest" in priorities:
            try:
                dist_val = float(h["dist"].replace("~", "").replace(" กม.", "").split()[0])
            except Exception:
                dist_val = 5.0
            scores["closest"] = max(0.0, (10.0 - dist_val) / 10.0) * 10
        if "value" in priorities:
            scores["value"] = max(0.0, (5000.0 - h["price"]) / 4000.0) * 10
        if "rated" in priorities:
            scores["rated"] = ((h["rating"] - 1.0) / 4.0) * 10
        if not scores:
            return 0.0
        return round(sum(scores.values()) / len(scores), 1)

    # sort: โรงแรมในงบก่อน แล้วค่อย rank ตาม score
    def sort_key(h: dict):
        within = h["price"] <= hotel_budget_per_night
        return (0 if within else 1, -hotel_score(h))

    sorted_hotels = sorted(hotels, key=sort_key)

    ranked_hotels = [
        RankedHotel(
            rank=i + 1,
            hotel=Hotel(**h),
            score=hotel_score(h),
            reasoning=_hotel_reasoning(h, priorities, h["price"] <= hotel_budget_per_night),
            withinBudget=h["price"] <= hotel_budget_per_night,
        )
        for i, h in enumerate(sorted_hotels)
    ]

    return RecommendResponse(rankedHotels=ranked_hotels)

File: frontend/lib/recommend.ts
import type { Hotel, RankedHotel, RecommendRequest, RecommendResponse } from "./types";
import { CONCERTS_DATA } from "./concerts";

const HOTEL_PRIORITY_LABELS: Record<string, string> = {
  closest: "ใกล้ฮอลล์ที่สุด",
  value: "คุ้มค่าที่สุด",
  rated: "รีวิวดีที่สุด",
};

function hotelReasoning(hotel: Hotel, priorities: string[], withinBudget: boolean): string {
  const frags: string[] = [];
  if (priorities.includes("closest")) frags.push(`อยู่ห่างจากฮอลล์แค่ ${hotel.dist}`);
  if (priorities.includes("value")) frags.push(`ราคาย่อมเยาที่ ฿${hotel.price.toLocaleString()}/คืน`);
  if (priorities.includes("rated")) frags.push(`คะแนนรีวิวสูงถึง ${hotel.rating}/5`);
  const labels = priorities
    .filter((p) => HOTEL_PRIORITY_LABELS[p])
    .map((p) => HOTEL_PRIORITY_LABELS[p])
    .join(" + ");
  const budgetNote = withinBudget ? "" : " (ราคาเกินงบที่เหลือ)";
  return `เพราะคุณเลือกเน้น "${labels}" — ${hotel.name} ตรงเงื่อนไขคุณ เพราะ${frags.join(" และ")} (${hotel.reason})${budgetNote}`;
}

function hotelScore(hotel: Hotel, priorities: string[]): number {
  const scores: number[] = [];
  if (priorities.includes("closest")) {
    const distVal = parseFloat(hotel.dist.replace("~", "").replace(" กม.", "").split(" ")[0]) || 5.0;
    scores.push(Math.max(0, (10.0 - distVal) / 10.0) * 10);
  }
  if (priorities.includes("value")) {
    scores.push(Math.max(0, (5000.0 - hotel.price) / 4000.0) * 10);
  }
  if (priorities.includes("rated")) {
    scores.push(((hotel.rating - 1.0) / 4.0) * 10);
  }
  if (scores.length === 0) return 0;
  return Math.round((scores.reduce((a, b) => a + b, 0) / scores.length) * 10) / 10;
}

export function getRecommendation(req: RecommendRequest): RecommendResponse {
  const concert = CONCERTS_DATA.find((c) => c.id === req.concertId);
  if (!concert) throw new Error("Concert not found");

  const priorities = req.hotelPriorities.length > 0 ? req.hotelPriorities : ["closest"];

  // คำนวณงบที่เหลือสำหรับโรงแรม
  const ticketPrice = concert.prices.find((p) => p.zone === req.preselectedZone)?.price ?? 0;
  const otherCosts = req.transportCost + req.merchCost + req.foodCost + req.otherCost;
  const hotelBudgetTotal = req.budget - ticketPrice - otherCosts;
  const hotelBudgetPerNight = req.hotelNights > 0 ? hotelBudgetTotal / req.hotelNights : hotelBudgetTotal;

  // sort: โรงแรมในงบก่อน แล้วค่อย rank ตาม score
  const sorted = [...concert.hotels].sort((a, b) => {
    const aInBudget = a.price <= hotelBudgetPerNight;
    const bInBudget = b.price <= hotelBudgetPerNight;
    if (aInBudget !== bInBudget) return aInBudget ? -1 : 1;
    return hotelScore(b, priorities) - hotelScore(a, priorities);
  });

  const rankedHotels: RankedHotel[] = sorted.map((hotel, i) => {
    const withinBudget = hotel.price <= hotelBudgetPerNight;
    return {
      rank: i + 1,
      hotel,
      score: hotelScore(hotel, priorities),
      reasoning: hotelReasoning(hotel, priorities, withinBudget),
      withinBudget,
    };
  });

  return { rankedHotels };
}
